src/Pos/Ssc/GodTossing/Toss/Logic.hs

```diff
@@ -105,13 +111,55 @@ rollbackGT oldestEOS (NewestFirst payloads)
 normalizeToss
     :: forall m . MonadToss m
     => EpochIndex -> TossModifier -> m ()
-normalizeToss epoch TossModifier{..} = do
-    putsUseful $ map (flip CommitmentsPayload mempty . mkCommitmentsMapUnsafe . one) $
-                 HM.toList $
-                 getCommitmentsMap _tmCommitments
-    putsUseful $ map (flip OpeningsPayload mempty . one) $ HM.toList _tmOpenings
-    putsUseful $ map (flip SharesPayload mempty . one) $ HM.toList _tmShares
-    putsUseful $ map (CertificatesPayload . one) $ HM.toList _tmCertificates
+normalizeToss epoch TossModifier {..} =
+    normalizeTossDo
+        epoch
+        ( HM.toList (getCommitmentsMap _tmCommitments)
+        , HM.toList _tmOpenings
+        , HM.toList _tmShares
+        , HM.toList _tmCertificates)
+
+-- | Apply the most valuable from given 'TossModifier' and drop the
+-- rest. This function can be used if mempool is exhausted.
+refreshToss
+    :: forall m . MonadToss m
+    => EpochIndex -> TossModifier -> m ()
+refreshToss epoch TossModifier {..} = do
+    comms <-
+        takeMostValuable epoch (HM.toList (getCommitmentsMap _tmCommitments))
+    opens <- takeMostValuable epoch (HM.toList _tmOpenings)
+    shares <- takeMostValuable epoch (HM.toList _tmShares)
+    certs <- takeMostValuable epoch (HM.toList _tmCertificates)
+    normalizeTossDo epoch (comms, opens, shares, certs)
+
+takeMostValuable ::
+       MonadToss m
+    => EpochIndex
+    -> [(StakeholderId, x)]
+    -> m [(StakeholderId, x)]
+takeMostValuable epoch items = take toTake <$> sortWithMDesc resolver items
+  where
+    toTake = 2 * length items `div` 3
+    resolver (id, _) =
+        fromMaybe (mkCoin 0) . (view (at id)) . fromMaybe mempty <$>
+        getRichmen epoch
```

Comment:
What if `getRichmen` changes over calls? I guess i'd retrieve it once and then use it for every `resolver` usage.